framework/src/media/MediaPlayerImpl.cpp

```diff
@@ -146,96 +222,194 @@ player_result_t MediaPlayerImpl::unprepare()
+		mpw.getPlayer()->pausePlayer();
+	}
+	mpw.setPlayer(shared_from_this());
+
+	mCurState = PLAYER_STATE_PLAYING;
+	notifyObserver(PLAYER_OBSERVER_COMMAND_STARTED);
+}
+
 player_result_t MediaPlayerImpl::stop()
 {
 	std::lock_guard<std::mutex> lock(mCmdMtx);
 	medvdbg("MediaPlayer stop\n");
-	if (mCurState == PLAYER_STATE_NONE || mCurState == PLAYER_STATE_IDLE) {
-		meddbg("MediaPlayer stop fail : wrong state\n");
+
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+	if (!mpw.isAlive()) {
 		return PLAYER_ERROR;
 	}
 
-	PlayerWorker& mpw = PlayerWorker::getWorker();
-	mpw.getQueue().enQueue(&PlayerWorker::stopPlayer, &mpw, shared_from_this());
+	mpw.getQueue().enQueue(&MediaPlayerImpl::stopPlayer, shared_from_this());
 
 	return PLAYER_OK;
 }
 
+void MediaPlayerImpl::stopPlayer()
+{
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+
+	medvdbg("MediaPlayer Worker : stopPlayer\n");
+	if (mCurState != PLAYER_STATE_PLAYING && mCurState != PLAYER_STATE_PAUSED) {
+		notifyObserver(PLAYER_OBSERVER_COMMAND_ERROR);
+		return;
+	}
+
+	mpw.setPlayer(nullptr);
+	mCurState = PLAYER_STATE_READY;
+}
+
 player_result_t MediaPlayerImpl::pause()
 {
 	std::lock_guard<std::mutex> lock(mCmdMtx);
 	medvdbg("MediaPlayer pause\n");
-	if (mCurState == PLAYER_STATE_NONE || mCurState == PLAYER_STATE_IDLE) {
-		meddbg("MediaPlayer pause fail : wrong state\n");
+
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+	if (!mpw.isAlive()) {
 		return PLAYER_ERROR;
 	}
 
-	PlayerWorker& mpw = PlayerWorker::getWorker();
-	mpw.getQueue().enQueue(&PlayerWorker::pausePlayer, &mpw, shared_from_this());
+	mpw.getQueue().enQueue(&MediaPlayerImpl::pausePlayer, shared_from_this());
 
 	return PLAYER_OK;
 }
 
+void MediaPlayerImpl::pausePlayer()
+{
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+
+	medvdbg("MediaPlayer Worker : pausePlayer\n");
+	if (mCurState != PLAYER_STATE_PLAYING) {
+		notifyObserver(PLAYER_OBSERVER_COMMAND_ERROR);
+		return;
+	}
+
+	mpw.setPlayer(nullptr);
+	mCurState = PLAYER_STATE_PAUSED;
+}
+
 int MediaPlayerImpl::getVolume()
 {
-	std::lock_guard<std::mutex> lock(mCmdMtx);
+	int ret = -1;
+
+	std::unique_lock<std::mutex> lock(mCmdMtx);
 	medvdbg("MediaPlayer getVolume\n");
+
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+	if (!mpw.isAlive()) {
+		return PLAYER_ERROR;
+	}
+
+	mpw.getQueue().enQueue(&MediaPlayerImpl::getVolumePlayer, shared_from_this(), std::ref(ret));
+	mSyncCv.wait(lock);
+
+	return ret;
+}
+
+void MediaPlayerImpl::getVolumePlayer(int &ret)
+{
+	medvdbg("MediaPlayer Worker : getVolume\n");
+
 	if (mCurState == PLAYER_STATE_NONE) {
 		meddbg("MediaPlayer getVolume fail : wrong state\n");
-		return -1;
+		notifyObserver(PLAYER_OBSERVER_COMMAND_ERROR);
+		ret = -1;
+		return;
 	}
 
-	return get_output_audio_volume();
+	ret = get_output_audio_volume();
+	mSyncCv.notify_one();
 }
 
 player_result_t MediaPlayerImpl::setVolume(int vol)
 {
-	std::lock_guard<std::mutex> lock(mCmdMtx);
+	player_result_t ret = PLAYER_ERROR;
```

Comment:
save as getVolume.

Reply:
getVolume returns int, and setVolume returns player_result_t. Why should setVolume save -1?

Reply:
I mean same comment as getVolume =), but we already discussed so nevermind.